### ISSUE TITLE
Filter by language doesn't work

### ISSUE THREAD

**Issue by mcarmonaa** [src-d/engine#277]
## Expected Behavior
Filtering by language should work properly

## Current Behavior
It doesn't apply the filter

## Steps to Reproduce (for bugs)
```scala
import tech.sourced.engine._

val engine = Engine(spark, "/path/to/siva-files", "siva")
val df = engine.getRepositories.getReferences.getCommits.getBlobs.classifyLanguages

df.where("lang='Python'").show

```

## Possible Solution
It seems that `spark` removes the filter as an optimization relying on the `Relation` to apply the filter. Since `lang` is a generated column by the `UDF`, this information isn't in the sources so the filter can't be applied at this level.

```
== Parsed Logical Plan ==
'Filter ('lang = Python)
+- Project [lang#280]
   +- Project [blob_id#58, commit_hash#59, repository_id#60, reference_name#61, content#62, is_binary#63, path#49, if (isnull(is_binary#63)) null else UDF(is_binary#63, path#49, content#62) AS lang#280]
      +- Project [blob_id#58, commit_hash#59, repository_id#60, reference_name#61, content#62, is_binary#63, path#49]
         +- Join Inner, (blob#50 = blob_id#58)
            :- SubqueryAlias blobs
            :  +- Relation[blob_id#58,commit_hash#59,repository_id#60,reference_name#61,content#62,is_binary#63] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(blob_id,StringType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(content,BinaryType,true), StructField(is_binary,BooleanType,false)),None,Some(blobs))
            +- Project [path#49, blob#50]
               +- Project [commit_hash#46, repository_id#47, reference_name#48, path#49, blob#50]
                  +- Join Inner, (commit_hash#46 = hash#21)
                     :- SubqueryAlias tree_entries
                     :  +- Relation[commit_hash#46,repository_id#47,reference_name#48,path#49,blob#50] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(path,StringType,false), StructField(blob,StringType,false)),None,Some(tree_entries))
                     +- Project [hash#21]
                        +- Project [repository_id#18, reference_name#19, index#20, hash#21, message#22, parents#23, parents_count#24, author_email#25, author_name#26, author_date#27, committer_email#28, committer_name#29, committer_date#30]
                           +- Project [repository_id#18, reference_name#19, index#20, hash#21, message#22, parents#23, parents_count#24, author_email#25, author_name#26, author_date#27, committer_email#28, committer_name#29, committer_date#30, repository_id#10]
                              +- Join Inner, ((repository_id#10 = repository_id#18) && (reference_name#19 = name#11))
                                 :- SubqueryAlias commits
                                 :  +- Relation[repository_id#18,reference_name#19,index#20,hash#21,message#22,parents#23,parents_count#24,author_email#25,author_name#26,author_date#27,committer_email#28,committer_name#29,committer_date#30] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(index,IntegerType,false), StructField(hash,StringType,false), StructField(message,StringType,false), StructField(parents,ArrayType(StringType,false),true), StructField(parents_count,IntegerType,false), StructField(author_email,StringType,true), StructField(author_name,StringType,true), StructField(author_date,TimestampType,true), StructField(committer_email,StringType,true), StructField(committer_name,StringType,true), StructField(committer_date,TimestampType,true)),None,Some(commits))
                                 +- Project [name#11, repository_id#10]
                                    +- Project [repository_id#10, name#11, hash#12]
                                       +- Join Inner, (repository_id#10 = id#0)
                                          :- SubqueryAlias references
                                          :  +- Relation[repository_id#10,name#11,hash#12] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(repository_id,StringType,false), StructField(name,StringType,false), StructField(hash,StringType,false)),None,Some(references))
                                          +- Project [id#0]
                                             +- SubqueryAlias repositories
                                                +- Relation[id#0,urls#1,is_fork#2,repository_path#3] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(id,StringType,false), StructField(urls,ArrayType(StringType,false),false), StructField(is_fork,BooleanType,true), StructField(repository_path,StringType,true)),None,Some(repositories))

== Analyzed Logical Plan ==
lang: string
Filter (lang#280 = Python)
+- Project [lang#280]
   +- Project [blob_id#58, commit_hash#59, repository_id#60, reference_name#61, content#62, is_binary#63, path#49, if (isnull(is_binary#63)) null else UDF(is_binary#63, path#49, content#62) AS lang#280]
      +- Project [blob_id#58, commit_hash#59, repository_id#60, reference_name#61, content#62, is_binary#63, path#49]
         +- Join Inner, (blob#50 = blob_id#58)
            :- SubqueryAlias blobs
            :  +- Relation[blob_id#58,commit_hash#59,repository_id#60,reference_name#61,content#62,is_binary#63] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(blob_id,StringType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(content,BinaryType,true), StructField(is_binary,BooleanType,false)),None,Some(blobs))
            +- Project [path#49, blob#50]
               +- Project [commit_hash#46, repository_id#47, reference_name#48, path#49, blob#50]
                  +- Join Inner, (commit_hash#46 = hash#21)
                     :- SubqueryAlias tree_entries
                     :  +- Relation[commit_hash#46,repository_id#47,reference_name#48,path#49,blob#50] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(path,StringType,false), StructField(blob,StringType,false)),None,Some(tree_entries))
                     +- Project [hash#21]
                        +- Project [repository_id#18, reference_name#19, index#20, hash#21, message#22, parents#23, parents_count#24, author_email#25, author_name#26, author_date#27, committer_email#28, committer_name#29, committer_date#30]
                           +- Project [repository_id#18, reference_name#19, index#20, hash#21, message#22, parents#23, parents_count#24, author_email#25, author_name#26, author_date#27, committer_email#28, committer_name#29, committer_date#30, repository_id#10]
                              +- Join Inner, ((repository_id#10 = repository_id#18) && (reference_name#19 = name#11))
                                 :- SubqueryAlias commits
                                 :  +- Relation[repository_id#18,reference_name#19,index#20,hash#21,message#22,parents#23,parents_count#24,author_email#25,author_name#26,author_date#27,committer_email#28,committer_name#29,committer_date#30] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(index,IntegerType,false), StructField(hash,StringType,false), StructField(message,StringType,false), StructField(parents,ArrayType(StringType,false),true), StructField(parents_count,IntegerType,false), StructField(author_email,StringType,true), StructField(author_name,StringType,true), StructField(author_date,TimestampType,true), StructField(committer_email,StringType,true), StructField(committer_name,StringType,true), StructField(committer_date,TimestampType,true)),None,Some(commits))
                                 +- Project [name#11, repository_id#10]
                                    +- Project [repository_id#10, name#11, hash#12]
                                       +- Join Inner, (repository_id#10 = id#0)
                                          :- SubqueryAlias references
                                          :  +- Relation[repository_id#10,name#11,hash#12] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(repository_id,StringType,false), StructField(name,StringType,false), StructField(hash,StringType,false)),None,Some(references))
                                          +- Project [id#0]
                                             +- SubqueryAlias repositories
                                                +- Relation[id#0,urls#1,is_fork#2,repository_path#3] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(id,StringType,false), StructField(urls,ArrayType(StringType,false),false), StructField(is_fork,BooleanType,true), StructField(repository_path,StringType,true)),None,Some(repositories))

== Optimized Logical Plan ==
Project [UDF(is_binary#63, path#49, content#62) AS lang#280]
+- Relation[blob_id#58,commit_hash#59,repository_id#60,reference_name#61,content#62,is_binary#63,commit_hash#46,repository_id#47,reference_name#48,path#49,blob#50,repository_id#18,reference_name#19,index#20,hash#21,message#22,parents#23,parents_count#24,author_email#25,author_name#26,author_date#27,committer_email#28,committer_name#29,committer_date#30,... 7 more fields] GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(blob_id,StringType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(content,BinaryType,true), StructField(is_binary,BooleanType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(path,StringType,false), StructField(blob,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(index,IntegerType,false), StructField(hash,StringType,false), StructField(message,StringType,false), StructField(parents,ArrayType(StringType,false),true), StructField(parents_count,IntegerType,false), StructField(author_email,StringType,true), StructField(author_name,StringType,true), StructField(author_date,TimestampType,true), StructField(committer_email,StringType,true), StructField(committer_name,StringType,true), StructField(committer_date,TimestampType,true), StructField(repository_id,StringType,false), StructField(name,StringType,false), StructField(hash,StringType,false), StructField(id,StringType,false), StructField(urls,ArrayType(StringType,false),false), StructField(is_fork,BooleanType,true), StructField(repository_path,StringType,true)),Some((((UDF(is_binary#63, path#49, content#62) = Python) && (blob#50 = blob_id#58)) && ((commit_hash#46 = hash#21) && (((repository_id#10 = repository_id#18) && (reference_name#19 = name#11)) && (repository_id#10 = id#0))))),None)

== Physical Plan ==
*Project [UDF(is_binary#63, path#49, content#62) AS lang#280]
+- *Scan GitRelation(org.apache.spark.sql.SparkSession@620ba2b0,StructType(StructField(blob_id,StringType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(content,BinaryType,true), StructField(is_binary,BooleanType,false), StructField(commit_hash,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(path,StringType,false), StructField(blob,StringType,false), StructField(repository_id,StringType,false), StructField(reference_name,StringType,false), StructField(index,IntegerType,false), StructField(hash,StringType,false), StructField(message,StringType,false), StructField(parents,ArrayType(StringType,false),true), StructField(parents_count,IntegerType,false), StructField(author_email,StringType,true), StructField(author_name,StringType,true), StructField(author_date,TimestampType,true), StructField(committer_email,StringType,true), StructField(committer_name,StringType,true), StructField(committer_date,TimestampType,true), StructField(repository_id,StringType,false), StructField(name,StringType,false), StructField(hash,StringType,false), StructField(id,StringType,false), StructField(urls,ArrayType(StringType,false),false), StructField(is_fork,BooleanType,true), StructField(repository_path,StringType,true)),Some((((UDF(is_binary#63, path#49, content#62) = Python) && (blob#50 = blob_id#58)) && ((commit_hash#46 = hash#21) && (((repository_id#10 = repository_id#18) && (reference_name#19 = name#11)) && (repository_id#10 = id#0))))),None) [is_binary#63,path#49,content#62] ReadSchema: struct<lang:string>
```

## Your Environment
* Version used: v0.3.0
* Operating System and version: 4.14.7-1-ARCH

**Comment by erizocosmico** [comment]
I recall that already happened sometime and we fixed that by collecting all non-redundant filters in all joins and adding them as filters.

**Comment by mcarmonaa** [comment]
Apparently fixed on #284. Pending of review.
